llama-cli/src/main.rs

```diff
@@ -35,12 +37,51 @@ fn main() {
         std::process::exit(1);
     };
 
-    let (model, vocab) = llama_rs::Model::load(&args.model_path, args.num_ctx_tokens as i32)
+    let (mut model, vocab) = llama_rs::Model::load(&args.model_path, args.num_ctx_tokens as i32)
         .expect("Could not load model");
 
+    if let Some(restore_path) = &args.restore_prompt {
+        let memory = ModelMemory::load_compressed(restore_path);
+        match memory.and_then(|memory| {
+            let res = model.set_memory(memory);
+            unsafe { model.get_memory(); }
+            res
+        }) {
+            Ok(_) => (),
+            Err(err) => {
+                eprintln!("Could not restore prompt. Error: {err}");
+                std::process::exit(1);
+            }
+        }
+    }
+
     let mut rng = thread_rng();
-    model.inference_with_prompt(&vocab, &inference_params, &prompt, &mut rng, |t| {
-        print!("{t}")
-    });
+    let stop_after_prompt = args.cache_prompt.is_some();
+    model.inference_with_prompt(
+        &vocab,
+        &inference_params,
+        &prompt,
+        &mut rng,
+        stop_after_prompt,
+        |t| {
+            print!("{t}");
+            std::io::stdout().flush().unwrap();
+        },
+    );
+
+    if let Some(cache_path) = &args.cache_prompt {
+        // SAFETY: no other model functions used inside the block
+        unsafe {
+            let memory = model.get_memory();
+            match memory.write_compressed(cache_path) {
```

Comment:
And this is how you access tensor memory and write it to a file. 

This returns a reference to tensor memory inside the inner context, so it's not safe to access that and then keep using the `model` object because that might mutate memory underneath and cause UB. I just decided to mark the function `unsafe` out of caution, but there's a whole underlying safety issue here that the `ggml` bindings allow mutating via a shared ref `&`. (more on that below)